File: template/solve.cpp
#include "utils.hpp"

#include <fmt/core.h>

#include <cassert>
#include <cstdlib>
#include <filesystem>
#include <string>
#include <vector>

namespace fs = std::filesystem;

// Custom input parser - modify based on problem
auto read_input(const fs::path &path) { return utils::read_lines(path); }

// Helper functions here

template <typename T>
auto part1(const T &input) {
    return 0;
}

template <typename T>
auto part2(const T &input) {
    return 0;
}

int main() {
    utils::set_working_dir(__FILE__);

    const auto example = read_input("./example.txt");
    const auto input = read_input("./input.txt");
    const auto part1_example_ans = 0;
    const auto part2_example_ans = 0;

    assert(part1(example) == part1_example_ans && "Part1 example failed");
    assert(part2(example) == part2_example_ans && "Part2 example failed");

    utils::Timer timer;
    auto p1 = part1(input);
    fmt::print("Part1: {} ({:.3f}ms)\n", p1, timer.elapsed_and_reset());

    auto p2 = part2(input);
    fmt::print("Part2: {} ({:.3f}ms)\n", p2, timer.elapsed());

    return 0;
}

File: template/solve.rs
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

fn set_working_dir() {
    let path = Path::new(file!());
    env::set_current_dir(path.parent().unwrap()).unwrap();
}

// Custom input parser - modify based on problem
fn read_input(path: &str) -> Vec<String> {
    fs::read_to_string(path)
        .expect(&format!("Failed to read {}", path))
        .lines()
        .map(|s| s.to_string())
        .collect()
}

// Helper functions here

fn part1<T>(input: &[T]) -> i64 {
    0
}

fn part2<T>(input: &[T]) -> i64 {
    0
}

fn main() {
    set_working_dir();

    let example = read_input("example.txt");
    let input = read_input("input.txt");
    let part1_example_ans = 0;
    let part2_example_ans = 0;

    assert_eq!(part1(&example), part1_example_ans, "Part1 example failed");
    assert_eq!(part2(&example), part2_example_ans, "Part2 example failed");

    let t1 = Instant::now();
    let p1 = part1(&input);
    let elapsed1 = t1.elapsed().as_secs_f64() * 1000.0;

    let t2 = Instant::now();
    let p2 = part2(&input);
    let elapsed2 = t2.elapsed().as_secs_f64() * 1000.0;

    println!("Part1: {} ({:.3}ms)", p1, elapsed1);
    println!("Part2: {} ({:.3}ms)", p2, elapsed2);
}
